### ISSUE TITLE
coredump when typing control char at yes/no prompt

### ISSUE THREAD

**Issue by dylnmc** [neovim/neovim#8070]
- `nvim --version`:
```
NVIM v0.2.2
Build type: Release
LuaJIT 2.0.5
Compilation: /usr/bin/cc -march=x86-64 -mtune=generic -O2 -pipe -fstack-protector-strong -fno-plt -Wconversion -DNVIM_MSGPACK_HAS_FLOAT32 -DNVIM_UNIBI_HAS_VAR_FROM -O2 -DNDEBUG -DMIN_LOG_LEVEL=3 -Wall -Wextra -pedantic -Wno-unused-parameter -Wstrict-prototypes -std=gnu99 -Wimplicit-fallthrough -Wvla -fstack-protector-strong -fdiagnostics-color=auto -DINCLUDE_GENERATED_DECLARATIONS -D_GNU_SOURCE -I/build/neovim/src/build/config -I/build/neovim/src/neovim-0.2.2/src -I/usr/include -I/usr/include -I/usr/include -I/usr/include -I/usr/include -I/usr/include -I/usr/include -I/build/neovim/src/build/src/nvim/auto -I/build/neovim/src/build/include
Compiled by builduser

Features: +acl +iconv +jemalloc +tui 
See ":help feature-compile"

   system vimrc file: "$VIM/sysinit.vim"
  fall-back for $VIM: "/usr/share/nvim"

Run :checkhealth for more info
```
- Operating system/version:
```
LSB_VERSION=1.4
DISTRIB_ID=Arch
DISTRIB_RELEASE=rolling
DISTRIB_DESCRIPTION="Arch Linux"
```
- Terminal name/version:
```
rxvt-unicode, version 9.22
```
- `$TERM`: `rxvt-unicode-256color`

### setup

not sure what caused problem, but here is screenshot

![screenshot from 2018-02-25 10-59-46](https://user-images.githubusercontent.com/7635158/36644445-a48eeeca-1a28-11e8-8eba-bbafcccfde51.png)

### problem

* seems like an issue with 'autoread'

* as you can see, I was in a split with `:windo diffthis` and had changed `init.vim` before trying to `:write`

```
           PID: 16872 (nvim)
           UID: 1000 (dylnmc)
           GID: 1000 (dylnmc)
        Signal: 6 (ABRT)
     Timestamp: Sun 2018-02-25 10:59:36 EST (54min ago)
  Command Line: nvim init.vim
    Executable: /usr/bin/nvim
 Control Group: /user.slice/user-1000.slice/session-c1.scope
          Unit: session-c1.scope
         Slice: user-1000.slice
       Session: c1
     Owner UID: 1000 (dylnmc)
       Boot ID: 27b19dab4a654bac99b9e8cd7c70563a
    Machine ID: c439fe12cf6240d698ea8a29da0fd363
      Hostname: xps
       Storage: /var/lib/systemd/coredump/core.nvim.1000.27b19dab4a654bac99b9e8cd7c70563a.16872.1519574376000000.lz4
       Message: Process 16872 (nvim) of user 1000 dumped core.
                
                Stack trace of thread 16872:
                #0  0x00007f63e0c1a860 raise (libc.so.6)
                #1  0x00007f63e0c1bec9 abort (libc.so.6)
                #2  0x00005621d4f03172 ui_puts (nvim)
                #3  0x00005621d4f031a8 ui_putc (nvim)
                #4  0x00005621d4ea2ba0 n/a (nvim)
                #5  0x00005621d4ea38c7 n/a (nvim)
                #6  0x00005621d4e32f44 n/a (nvim)
                #7  0x00005621d4e35014 n/a (nvim)
                #8  0x00005621d4e35830 msg_puts_attr_len (nvim)
                #9  0x00005621d4e359ff msg_putchar_attr (nvim)
                #10 0x00005621d4e3cb7b ask_yesno (nvim)
                #11 0x00005621d4df827f buf_write (nvim)
                #12 0x00005621d4db6385 do_write (nvim)
                #13 0x00005621d4dd1fb4 do_cmdline (nvim)
                #14 0x00005621d4e4ba38 n/a (nvim)
                #15 0x00005621d4e54f98 n/a (nvim)
                #16 0x00005621d4ee3b16 state_enter (nvim)
                #17 0x00005621d4e4c3ea normal_enter (nvim)
                #18 0x00005621d4d3605a main (nvim)
                #19 0x00007f63e0c06f4a __libc_start_main (libc.so.6)
                #20 0x00005621d4d36eea _start (nvim)
                
                Stack trace of thread 16875:
                #0  0x00005621d4db03b8 multiqueue_empty (nvim)
                #1  0x00005621d4daf746 n/a (nvim)
                #2  0x00007f63e26396c4 n/a (libuv.so.1)
                #3  0x00007f63e2648f00 uv__io_poll (libuv.so.1)
                #4  0x00007f63e2639ff3 uv_run (libuv.so.1)
                #5  0x00005621d4daf9e2 loop_poll_events (nvim)
                #6  0x00005621d4f00463 n/a (nvim)
                #7  0x00007f63e241608c start_thread (libpthread.so.0)
                #8  0x00007f63e0cdbe7f __clone (libc.so.6)
```

### config

``` vim
if has("win32")
    let $MYVIMDIR=$HOME.'/nvimfiles'
else
    if has("unix")
        let s:uname = system("uname")
        if s:uname == "Darwin\n"
            let $MYVIMDIR=$HOME.'/.config/nvim'
        elseif s:uname == "Linux\n"
            let $MYVIMDIR=$HOME.'/.config/nvim'
        endif
    endif
endif

scriptencoding utf-8

" search through all subdirs with many commands
set path+=**

" cmd line completion on wildchar
set wildmenu

" make backspace work all the time
set backspace=indent,eol,start

" automatic smart indenting
set smartindent

" breakindent+showbreak character for visibility of broken lines
set breakindent
let &showbreak='↪'

" ft detection for language-dependent indenting
filetype plugin indent on

" create undo, backup, tmp dirs for vim if they don't exist
if !isdirectory($MYVIMDIR.'/undo')   | call mkdir($MYVIMDIR.'/undo', 'p')   | endif
if !isdirectory($MYVIMDIR.'/backup') | call mkdir($MYVIMDIR.'/backup', 'p') | endif
if !isdirectory($MYVIMDIR.'/tmp')    | call mkdir($MYVIMDIR.'/tmp', 'p')    | endif

" don't clutter current dir but use awesome functionality that vim provides
" undo
set undodir=$MYVIMDIR/undo
set undofile
" backup
set backupdir=$MYVIMDIR/backup
set backup
" swap
set directory=$MYVIMDIR/tmp
set swapfile
if (!has('win32') && !has('nvim'))
    " viminfo
    set viminfofile=$MYVIMDIR/viminfo
endif

" numbering
set number
set relativenumber

" buffers become hidden when abandoned
set hidden

" auto-read/auto-write
set autoread
set autowrite

" no bells whatsoever
set noerrorbells | set novisualbell | set t_vb= | set belloff=all

" completion from current and included files
set complete+=i

" all tabs are 4 spaces visually and literally!
set tabstop=4
set shiftwidth=4
set softtabstop=4

" spaces instead of tab
set expandtab

augroup make
    autocmd!
    autocmd FileType make,cmake setlocal noexpandtab
augroup end

" round indent to multiple of shiftwidth
set shiftround

" incremental search and highlight searches
set incsearch
set hlsearch
" <c-l> is a bit more useful (nohl && diffu && <c-l>)
nnoremap <silent> <c-l> :nohlsearch<bar>diffupdate<cr><c-l>

" don't autowrap lines
set textwidth=0

" prevent warnings and annoyances
set shortmess=aAIqFcTWsqt

" smart-case search
set ignorecase
set smartcase

" split below/right
set splitbelow
set splitright

" tilde + op
set tildeop

" show cmd when typing
set showcmd

" show mode
set noshowmode

" visual block
set virtualedit=block,insert

" omni-completion <c-x><c-o>
set omnifunc=syntaxcomplete#Complete

" cursorline
set nocursorline

" colorcolumn
set colorcolumn=

" (max) history
set history=10000

" dvorak in insert
" set keymap=dvorak

" list chars
set list
set listchars=tab:»\·,trail:·,nbsp:~

" delete comment char when joining lines
set formatoptions+=j

" fewer "press enter" prompts
set cmdheight=2

" diff opens in vertical split
set diffopt+=vertical

" keep 1 line above/below cursor at top/bottom of screen (resp.)
set scrolloff=1

" don't add two spaces after punctuation
set nojoinspaces

" ultisnips
let g:UltiSnipsExpandTrigger            = "<tab>"
let g:UltiSnipsListSnippets             = "<c-l>"
let g:UltiSnipsJumpForwardTrigger       = "<c-j>"
let g:UltiSnipsJumpBackwardTrigger      = "<c-k>"
let g:UltiSnipsSnippetsDir              = $MYVIMDIR.'/plugged/vim-ultisnippets/UltiSnips/'
let g:UltiSnipsSnippetDirectories       = [ "UltiSnips" ]
let g:ultisnips_python_quoting_style    = "single"

" prevent normal mode completions for emmet (<c-y> messes up 50% of the time)
let g:user_emmet_mode = 'iv'

let g:javascript_plugin_jsdoc      = 1
let g:jsdoc_allow_input_prompt     = 1

let g:nord_italic_comments         = 1
let g:nord_uniform_diff_background = 1
" let g:nord_uniform_status_lines    = 1
let g:nord_comment_brightness = 4

let g:airline_powerline_fonts = 1 
let g:airline#extensions#whitespace#enabled = 0

" stop lag for statusline
set ttimeoutlen=50

" plug
if empty(glob($MYVIMDIR.'/autoload/plug.vim'))
    silent !curl -fLo $MYVIMDIR/autoload/plug.vim --create-dirs https://raw.githubusercontent.com/junegunn/vim-plug/master/plug.vim
    autocmd VimEnter * PlugInstall --sync | source $MYVIMRC
endif
call plug#begin($MYVIMDIR.'/plugged')
" favs
Plug 'SirVer/ultisnips'
Plug 'tomtom/tcomment_vim'
Plug 'tpope/vim-repeat'

" use fairly freq
Plug 'tpope/vim-surround'
Plug 'dylnmc/vim-ultisnippets'
Plug 'docunext/closetag.vim'
Plug 'kana/vim-arpeggio'
Plug 'vim-scripts/ReplaceWithRegister'
Plug 'tpope/vim-unimpaired'
Plug 'tpope/vim-fugitive'
Plug 'tommcdo/vim-exchange'
Plug 'dylnmc/vim-logipair'
Plug 'godlygeek/tabular'
Plug 'sickill/vim-pasta'

" testing
Plug 'mbbill/undotree'
" Plug 'simnalamburt/vim-mundo'
Plug 'lervag/vimtex'
Plug 'dylnmc/vim-latex-live-preview', { 'for': 'tex' }

" color/configuration
Plug 'dylnmc/lyne-vim'
Plug 'arcticicestudio/nord-vim', { 'branch': 'develop' }
Plug 'morhetz/gruvbox' " backup colo
call plug#end()

" Load matchit.vim
runtime! macros/matchit.vim

let g:lyne_options = {
    \'modify_color_left_num': 2,
    \'modify_color_right_num': 2,
    \'helpMode': ''
\}

if has('win32')
    call extend(g:lyne_options, {
        \'leftSeparator': '|',
        \'rightSeparator': '|',
        \'leftDivider': '»',
        \'rightDivider': '«'
    \})
endif

function! g:LyneGit(...)
    let l:active = a:1 == winnr()
    if (l:active && exists('*fugitive#head'))
        let l:branch = fugitive#head()
        " let l:branch = getbufvar(winbufnr(a:winnum), 'call fugitive#statusline()')
        return l:branch ==# '' ? '' : ' '.l:branch
    endif
    return ''
endfunction

let g:lyne_segments = {
    \'left':  [ ['lyne#getMode'], ['g:LyneGit', 'lyne#getName', 'lyne#getModified'], ['lyne#getFiletype'] ],
    \'right': [ ['lyne#getColNumber'], [ 'lyne#getLineNumber' ] ]
\}

augroup LyneColors
    autocmd!
    " autocmd ColorScheme * highlight LyneModeNormal ctermbg=7 ctermfg=8 guibg=white guifg=#2e2e2e
    autocmd ColorScheme * highlight StatusLine guibg=#4C566B

    autocmd VimEnter,ColorScheme nord highlight LyneModeNormal  ctermfg=8 ctermbg=4 guifg=#4C566A guibg=#81A1C1
    autocmd VimEnter,ColorScheme nord highlight LyneModeInsert  ctermfg=8 ctermbg=7 guifg=#4C566A guibg=#ECEFF4
    autocmd VimEnter,ColorScheme nord highlight LyneModeVisual  ctermfg=8 ctermbg=5 guifg=#4C566A guibg=#B48EAD
    autocmd VimEnter,ColorScheme nord highlight LyneModeReplace ctermfg=8 ctermbg=1 guifg=#4C566A guibg=#BF616A
    " Left
    autocmd VimEnter,ColorScheme nord highlight LyneLeft1 ctermfg=NONE ctermbg=NONE guifg=NONE guibg=NONE
    autocmd VimEnter,ColorScheme nord highlight LyneLeft2 ctermfg=7 ctermbg=0 guifg=#E5E9F0 guibg=#3B4252
    autocmd VimEnter,ColorScheme nord highlight LyneLeft3 ctermfg=7 ctermbg=8 guifg=#E5E9F0 guibg=#4C566A
    " " Right
    autocmd VimEnter,ColorScheme nord highlight LyneRight1 ctermfg=8 ctermbg=4 guifg=#3B4252 guibg=#81A1C1
    autocmd VimEnter,ColorScheme nord highlight LyneRight2 ctermfg=7 ctermbg=0 guifg=#E5E9F0 guibg=#3B4252
    " " Modified
    autocmd VimEnter,ColorScheme nord highlight LyneModified ctermfg=6 ctermbg=0 guifg=#88C0D0 guibg=#3B4252
augroup end

" color
augroup nord
    autocmd!
    autocmd ColorScheme nord highlight shDerefSimple ctermfg=6 guifg=#88C0D0
    autocmd ColorScheme nord highlight shDerefVar ctermfg=6 guifg=#88C0D0
    autocmd ColorScheme nord highlight shVariable ctermfg=6 guifg=#88C0D0

    autocmd ColorScheme nord highlight StatusLineNC ctermbg=8 guibg=#4C566A ctermfg=14 guifg=#88C0D0
    autocmd ColorScheme nord highlight Conceal ctermbg=NONE ctermfg=6 guibg=NONE guifg=#88C0D0 cterm=italic,bold gui=italic,bold
augroup end

" adhere to nostatusline wednesday
" if (strftime('%w') == 3)
"     set laststatus=0
" else
    set laststatus=2
" endif

syntax enable
set background=dark
colorscheme nord
if &term =~ '256color' && exists('$TMUX')
    set t_ut=
endif

" <space> toggle folding
nnoremap <space> za
vnoremap <space> za

" netrw plugin configuration
let g:netrw_banner = 0

" restore cursor position when reopening a file
augroup curposRestore
    autocmd!
    autocmd BufReadPost * call setpos(".", getpos("'\""))
augroup end

" *don't* insert comment when using o/O in normal mode
" *do* insert comment when pressing <cr> in insert mode
augroup stopCommentTsunami
    autocmd!
    autocmd FileType * setlocal formatoptions+=c
    autocmd FileType * setlocal formatoptions-=o
augroup end

if exists("*synstack")
    " Show syntax highlighting groups for word under cursor
    nnoremap <c-p> :call <sid>SynStack()<cr>
    function! <sid>SynStack()
        echo map(synstack(line('.'), col('.')), 'synIDattr(v:val, "name")')
    endfunc
endif

" I don't like mouse
set mouse=

try
    " easy file saving (press 'jk' at--almost--same time)
    call arpeggio#map('n', '', 0, 'jk', ':silent write<cr>:<c-c>')
    call arpeggio#map('i', '', 0, 'jk', '<esc>:silent write<cr>:<c-c>')
catch
endtry

set wildcharm=<c-z>
nnoremap <leader>b :buffers<cr>:b<space>

nnoremap * /\<<c-r><c-w>\><cr>
nnoremap # ?\<<c-r><c-w>\><cr>
nnoremap g* /<c-r><c-w><cr>
nnoremap g# ?<c-r><c-w><cr>

" I use vim as my man page viewer
augroup man
    autocmd!
    autocmd VimEnter,BufWinEnter *.~ nnoremap <c-h> :execute "normal! `Z"<cr>
    autocmd VimEnter,BufWinEnter *.~ nnoremap <buffer> K :execute "normal! mZ:Man \<lt>c-r>\<lt>c-w>\<lt>cr>"<cr>
    autocmd VimEnter,BufWinEnter *.~ vnoremap <buffer> K ymZ:Man <c-r>"<cr>"<cr>
    autocmd VimEnter,BufWinEnter *.~ set number relativenumber
augroup end

autocmd BufReadPost,BufNewFile *.otl set filetype=otl
```

**Comment by jamessan** [comment]
The crash was caused by us `abort()`ing in `ui_puts`:

https://github.com/neovim/neovim/blob/158f8b7ce3051127918148f7f8e9892a4c9ae34f/src/nvim/ui.c#L418-L420

Not sure what we were trying to "put", though.

**Comment by justinmk** [comment]
Thanks for the report. Can you try the development version? The [Releases](https://github.com/neovim/neovim/releases) page has pre-built archives for Linux/Windows/macOS. (No installation needed, just unzip and run. For Linux, [nvim.appimage](https://github.com/neovim/neovim/releases/download/nightly/nvim.appimage) is a universal "binary", just `chmod u+x nvim.appimage && ./nvim.appimage`)

**Comment by dylnmc** [comment]
Perhaps I wasn't clear; I am not quite able to replicate it (not great for debugging on your end), but I thought surely something that you would want to take a look at given it's a seg fault.

I just tried again in my installed version of nvim and was unable to replicate the issue. I tried quite hard to, as well :disappointed:

**Comment by bfredl** [comment]
I managed to reproduce it: at the yesno prompt type some ctrl char, like `<C-E>`. (if you didn't type it, it could've come from a mapping or something). `message.c` needs to check the char is printable before sending it to `screen.c` .

**Comment by dylnmc** [comment]
@bfredl is right! all you need to do is as follows:

#### session 1
```
nvim<cr>
:e $MYVIMRC<cr>
```

#### session 2
```
nvim<cr>
:e $MYVIMRC<cr>
Go" random comment<esc>
:w<cr>
```

#### session 1 (again)

```
:w<cr>
<c-e>
```

then, you get a seg fault.

aside: I extracted the zip for the dev version for linux, and the same issue occurred.

Cheers,
dylnmc

**Comment by justinmk** [comment]
This is the same as https://github.com/neovim/neovim/issues/7586 where I bisected to https://github.com/neovim/neovim/commit/7d6af9985c057eef8f6bb2a625744df1b4ad0a72

**Comment by bfredl** [comment]
fixed by #8070
